Removing the auto-apply feature from Time Series Visual Builder

simianhacker opened the pull request:
This PR changes the auto apply feature in Time Series Visual Builder and adds a button to the interface to apply changes when the user has made a change (when auto apply is disabled).  This also adds a toggle to turn on or off the auto apply feature. 

With auto apply on: 

![image](https://cloud.githubusercontent.com/assets/41702/25826216/d86979a0-33fa-11e7-954c-bec724ea10b7.png)

With auto apply off:

![image](https://cloud.githubusercontent.com/assets/41702/25826256/fdea80f2-33fa-11e7-96eb-1a0fbaa2152f.png)


CC @trevan

thomasneirynck commented on src/core_plugins/metrics/public/kbn_vis_types/editor_controller.js:
```diff
@@ -28,12 +28,12 @@ app.controller('MetricsEditorController', (
+  // }, 1000, {
+  //   leading: false,
+  //   trailing: true
+  // });
```
remove

simianhacker commented:
@alexf OR @tbragin... Can either of you weigh in on this change? This changes the behavior and addresses issues brought up in a separate discussion  https://github.com/elastic/kibana/pull/11094#issuecomment-292570439

tbragin commented:
@simianhacker I played with this PR and I am not a fan of hiding the "Apply" button when no changes are detected. I'd recommend following the same approach that Visualize uses, and disabling it on no changes.

simianhacker commented:
I just updated this PR to change the interface. Now the apply bar is always visible. I also added a toggle to turn on and off the auto apply feature. Finally I changed the way the dirty checking is applied, it now only applies if something in the series has changed.

simianhacker commented:
I also updated the description with new images that reflect this change.

simianhacker commented:
jenkins test this

simianhacker commented:
@thomasneirynck Do you want to give this another review? I changed the behavior a little.

ppisljar commented:
not sure if the toggle button should be saved with visualization, or should it be remembered per user ? any case i am ok with moving forward as it is.

i vote for moving toggle and apply close together as well.

simianhacker commented:
I like the idea of saving the toggle state with the user (localStorage). That seems to make the most sense to me since it's a interface preference.

simianhacker commented:
I moved the toggle and the button to be side by side:

![image](https://cloud.githubusercontent.com/assets/41702/26133121/ea162634-3a72-11e7-9cd5-11f2b70482d9.png)

@ppisljar @thomasneirynck I also saved the auto apply state to `localStorage` so the setting would stay with the user.

ppisljar commented:
visually its not the most pleasing (buttons together) ... but they are easier to find. i am ok with any way, whichever you prefer. maybe ask design team for suggestion if you feel its worth it ?

simianhacker commented:
Back ported to 5.x with b48b1ee
